src/paho/mqtt/client.py

```diff
@@ -2176,19 +2420,28 @@ def on_socket_open(self, func):
         with self._callback_mutex:
             self._on_socket_open = func
 
-    def socket_open_callback(self):
-        def decorator(func):
+    def socket_open_callback(
+        self,
+    ) -> typing.Callable[[CallbackOnSocket], CallbackOnSocket]:
+        def decorator(func: CallbackOnSocket) -> CallbackOnSocket:
             self.on_socket_open = func
             return func
         return decorator
 
-    def _call_socket_open(self):
+    def _call_socket_open(self) -> None:
         """Call the socket_open callback with the just-opened socket"""
         with self._callback_mutex:
             on_socket_open = self.on_socket_open
 
         if on_socket_open:
             with self._in_callback_mutex:
+                if self._sock is None:
+                    self._easy_log(
+                        MQTT_LOG_ERR,
+                        "socket() is None in _call_socket_open",
+                    )
+                    return
```

Comment:
This sounds like a "should-never-happen" sort of bug – maybe `_call_socket_open()`should be called by the client internally with the newly opened socket as an argument, so it wouldn't need `self._sock`?